File: src/fkt/core/IndexingFrameTargeter.java
package fkt.core;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import fkt.util.Debug;
public class IndexingFrameTargeter extends TargeterCore{
  final private Map<String,STargeter>titleTargeters=new HashMap();
	private STargeter indexing,indexed;
	private SIndexing indexingTarget;
	private STarget indexedTarget;
	private String indexedTitle;
	public void retarget(STarget target){
		super.retarget(target);
		updateToTarget();
		if(indexing==null){
	  	indexing=indexingTarget.newTargeter();
	  	indexing.setNotifiable(this);
	  }
		if(titleTargeters.isEmpty()){
			int atThen=indexingTarget.index();
			for(int at=0;at<indexingTarget.indexables().length;at++){
				indexingTarget.setIndex(at);
				updateToTarget();
				indexed=((TargetCore)indexedTarget).newTargeter();
				indexed.setNotifiable(this); 
				indexed.retarget(indexedTarget);
				titleTargeters.put(indexedTitle,indexed);
			}
			indexingTarget.setIndex(atThen);
			updateToTarget();
		}
		indexing.retarget(indexingTarget);
		indexed=titleTargeters.get(indexedTitle);
		if(indexed==null)throw new IllegalStateException("Null indexed in "+this);
	  indexed.retarget(indexedTarget);
	}
	private void updateToTarget(){
		IndexingFrame frame=(IndexingFrame)target();
		indexingTarget=frame.indexing();
		indexedTarget=frame.indexedTarget();
		indexedTitle=indexedTarget.title();
	}
  public void retargetFacets(){
    super.retargetFacets();
    indexing.retargetFacets();
    for(STargeter t:titleTargeters.values())t.retargetFacets();
  }
	public STargeter[]titleElements(){
  	ArrayList<STargeter>list=new ArrayList<STargeter>(Arrays.asList(elements));
		list.add(indexing);
		for(STargeter t:titleTargeters.values())list.add(t);
		return list.toArray(new STargeter[]{});
	}
}

File: src/fkt/facets/core/IndexingFrameTargeter.kt
package fkt.facets.core

class IndexingFrameTargeter : TargeterCore("IndexingFrameTargeter") {
  private val titleTargeters = HashMap<String, Targeter>()
  private lateinit var indexing: Targeter
  private lateinit var indexed: Targeter
  private lateinit var indexingTarget: Indexing
  private lateinit var indexedTarget: Targety
  private lateinit var indexedTitle: String
  override fun retarget(target: Targety) {
    super.retarget(target)
    updateToTarget()
    if(!::indexing.isInitialized){
      indexing = indexingTarget.newTargeter()
      if(false)trace(".retarget: indexing=",indexing)
    }
    indexing.setNotifiable(this)
    if (titleTargeters.size == 0) {
      val atThen = indexingTarget.index
      for (at in indexingTarget.indexables().indices) {
        indexingTarget.index=at
        updateToTarget()
        if(false)trace(".retarget: indexedTarget=",indexedTarget)
        indexed = (indexedTarget as TargetCore).newTargeter()
        val indexed = indexed
        indexed.setNotifiable(this)
        indexed.retarget(indexedTarget)
        titleTargeters[indexedTitle] = indexed
      }
      indexingTarget.index=atThen
      updateToTarget()
    }
    indexing.retarget(indexingTarget)
    indexed = titleTargeters[indexedTitle] as Targeter
    indexed.retarget(indexedTarget)
  }

  override fun retargetFacets() {
    super.retargetFacets()
    indexing.retargetFacets()
    titleTargeters.values.forEach { it.retargetFacets()}
  }

  override val titleElements: List<Targeter> get() {
    val list = elements.toMutableList()
    list.add(indexing)
    titleTargeters.values.forEach{ it ->list.add(it)}
    return list
  }

  private fun updateToTarget() {
    val frame = target as IndexingFrame
    indexingTarget = frame.indexing()
    indexedTarget = frame.indexedTarget()
    indexedTitle = indexedTarget.title
  }
}

